OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs

```diff
@@ -67,20 +74,17 @@ class UpgradeState
 		}
 
 		readonly List<TimedUpgrade> timedUpgrades = new List<TimedUpgrade>();
-		readonly Lazy<Dictionary<string, UpgradeState>> upgrades;
 		readonly Dictionary<IUpgradable, int> levels = new Dictionary<IUpgradable, int>();
+		Dictionary<string, UpgradeState> upgrades;
 
-		public UpgradeManager(ActorInitializer init)
-		{
-			upgrades = Exts.Lazy(() =>
-			{
-				var ret = new Dictionary<string, UpgradeState>();
-				foreach (var up in init.Self.TraitsImplementing<IUpgradable>())
-					foreach (var t in up.UpgradeTypes)
-						ret.GetOrAdd(t).Traits.Add(up);
+		public UpgradeManager(ActorInitializer init) { }
 
-				return ret;
-			});
+		public void Created(Actor self)
```

Comment:
Since you are introducing this here and don't need to call this yourself, please explicitly implement this.